### PR TITLE
feat: enable hermetic library generation

### PR DESCRIPTION
**Pull request by diegomarquezp**
This is part of the effort to enable multi-version support in hand-written libraries. More context and details in [this doc](https://docs.google.com/document/d/1wrpyBtphdenM3BNelcnpBKGADYrGJUo686HXvSA0h-0/edit?pli=1&tab=t.0#heading=h.lbr9t8ls3oip).

This PR adds two workflows and their scripts:
 * `update_generation_config` which updates `generation_config.yaml` (added in this PR) to contain the latest `googleapis_committish` (source of truth of library generation) as well as other dependencies such as `gapic_generator_version`. This job runs nightly and creates a PR that modifies the config yaml.
 * `hermetic_library_generation` which reads `generation_config.yaml` and pushes a new commit to any PR whose latest commit affects `generation_config.yaml` (e.g. the nightly PR created by `update_generation_config`). It essentially translates changes in the config yaml to changes in the source code (of the generated part). This workflow will be a required check.

As an example of how they work in practice:
 1. Once per night, `update_generation_config` creates a PR that updates `generation_config.yaml` with the latest dependencies and googleapis committish
 2. `hermetic_library_generation` is triggered by this PR and generates the corresponding code changes. A new commit with the code changes is added to the PR. This step may occur for any PR affecting `generation_config.yaml`.

### Follow ups
 - [ ] Add `CLOUD_JAVA_BOT_TOKEN` to repository for the hermetic generation workflow to work

### CODE REVIEW

File: generation/hermetic_library_generation.sh [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+fi
+# copy generation configuration from target branch to current branch.
+git show "${target_branch}":"${generation_config}" > "${baseline_generation_config}"
+config_diff=$(diff "${generation_config}" "${baseline_generation_config}" || true)
+
+# run hermetic code generation docker image.
+docker run \
+  --rm \
+  -u "$(id -u):$(id -g)" \
+  -v "$(pwd):${workspace_name}" \
```
Comment (JoeWang1127) [inline generation/hermetic_library_generation.sh:95]: Can you add `-v "$HOME"/.m2:/home/.m2` so that this script can also used to generate common protos?

**Review by JoeWang1127 — APPROVED** [review]

File: .github/workflows/hermetic_library_generation.yaml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    - name: Generate changed libraries
+      shell: bash
+      run: |
+        set -x
+        [ -z "$(git config user.email)" ] && git config --global user.email "cloud-java-bot@google.com"
+        [ -z "$(git config user.name)" ] && git config --global user.name "cloud-java-bot"
+        bash generation/hermetic_library_generation.sh \
+          --target_branch ${{ github.base_ref }} \
+          --current_branch ${{ github.head_ref }} \
+          --image_tag $(cat generation_config.yaml | yq .gapic_generator_version)
```
Comment (JoeWang1127) [inline .github/workflows/hermetic_library_generation.yaml:37]: Is `yg` available in ubuntu by default?

Comment (diegomarquezp) [reply .github/workflows/hermetic_library_generation.yaml:37]: Yes, [this is the test](https://github.com/diegomarquezp/java-bigtable/actions/runs/9213821854/job/25348574274?pr=11)

File: generation/hermetic_library_generation.sh (hunk 1 above)
Comment (diegomarquezp) [reply generation/hermetic_library_generation.sh:95]: Done

File: generation/hermetic_library_generation.sh [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+  --baseline-generation-config-path="${workspace_name}/${baseline_generation_config}" \
+  --current-generation-config-path="${workspace_name}/${generation_config}"
+
+
+# commit the change to the pull request.
+if [[ $(basename $(pwd)) == "google-cloud-java" ]]; then
+  git add java-* pom.xml gapic-libraries-bom/pom.xml versions.txt
+else
+  # The image leaves intermediate folders and files it works with. Here we remove them
+  rm -rdf output googleapis baseline_generation_config.yaml pr_description.txt
```
Comment (JoeWang1127) [inline generation/hermetic_library_generation.sh:106]: why do we need to remove pr_description.txt?

Comment (diegomarquezp) [reply generation/hermetic_library_generation.sh:106]: It's part of the cleanup. A blind `git add .` would include it otherwise

Comment (diegomarquezp) [reply generation/hermetic_library_generation.sh]: That makes sense. I modified it as suggested. Thanks!

File: generation_config.yaml [hunk 4]
```diff
@@ -0,0 +1,37 @@
+protoc_version: '25.3'
```
[the remark this answers is not in the document]
Comment (blakeli0) [reply generation_config.yaml:1]: Actually protoc version is now baked into our [docker image](https://github.com/googleapis/sdk-platform-java/blob/main/.cloudbuild/library_generation/library_generation.Dockerfile#L21), @diegomarquezp can you please remove it to reflect the recent changes? 

The protoc version used to come from the [googleapis/WORKSPACE](https://github.com/googleapis/googleapis/blob/master/WORKSPACE#L152), which is shared by all languages. This is now configurable by Java.

File: .github/scripts/update_generation_config.sh [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+# This script should be run at the root of the repository.
+# This script is used to update googleapis_commitish, gapic_generator_version,
+# and libraries_bom_version in generation configuration at the time of running
+# and create a pull request.
+
+# The following commands need to be installed before running the script:
+# 1. git
+# 2. gh
+# 3. jq
+
```
[the remark this answers is not in the document]
Comment (diegomarquezp) [reply .github/scripts/update_generation_config.sh:12]: Done

File: generation_config.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (diegomarquezp) [reply generation_config.yaml]: Done

File: .github/workflows/update_generation_config.yaml [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+      # the branch into which the pull request is merged
+      base_branch: main
+    steps:
+    - uses: actions/checkout@v4
+      with:
+        token: ${{ secrets.CLOUD_JAVA_BOT_TOKEN }}
+    - name: Update params in generation config to latest
+      shell: bash
+      run: |
+        set -x
```
[the remark this answers is not in the document]
Comment (diegomarquezp) [reply .github/workflows/update_generation_config.yaml]: Added -e

File: .github/workflows/hermetic_library_generation.yaml [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v4
+      with:
+        fetch-depth: 0
+        token: ${{ secrets.CLOUD_JAVA_BOT_TOKEN }}
+    - name: Generate changed libraries
+      shell: bash
+      run: |
+        set -x
```
[the remark this answers is not in the document]
Comment (diegomarquezp) [reply .github/workflows/hermetic_library_generation.yaml]: Added -e

**Pull request merged**